src/assign.c

```diff
@@ -435,7 +435,10 @@ SEXP assign(SEXP dt, SEXP rows, SEXP cols, SEXP newcolnames, SEXP values, SEXP v
   if (length(newcolnames)) {
     oldtncol = TRUELENGTH(dt);   // TO DO: oldtncol can be just called tl now, as we won't realloc here any more.
 
-    if (oldtncol<oldncol) error("Internal error, please report (including result of sessionInfo()) to data.table issue tracker: oldtncol (%d) < oldncol (%d) but tl of class is marked.", oldtncol, oldncol); // # nocov
+    if (oldtncol<oldncol) {
+      // Likely user has loaded this object and not run setDT on it, e.g. #2996
+      if (oldtncol == 0) error("It appears this object has not yet had any columns; if you've loaded this data.table from disk, e.g. with load(), please remember to always call setDT() after loading to allocate columns. If the issue persists, please report (including result of sessionInfo()) to data.table issue tracker: oldtncol (%d) < oldncol (%d) but tl of class is marked.", oldtncol, oldncol); // # nocov
```

Comment:
This one could be tested with `saveRDS`, `readRDS`, then `nocov` can be removed
I would also mention `readRDS` together with `load`, which in most cases should be preferred over `load`.
And phrasing `allocate columns` -> `pre-allocate space for new columns`?